actionpack/lib/action_dispatch/routing/url_for.rb

```diff
@@ -180,7 +180,8 @@ def url_for(options = nil)
         when Symbol
           HelperMethodBuilder.url.handle_string_call self, options
         when Array
-          polymorphic_url(options, options.extract_options!)
+          components = options.dup
+          polymorphic_url(components, components.extract_options!)
```

Comment:
Can we fix `polymorphic_url` to not change the object instead of calling `dup` where it is called?